### ISSUE TITLE
Need to hack `urlify` to support custom types without writing new loader

### ISSUE THREAD

**Issue by ezralalonde** [MoOx/phenomic#852]
Hi again,

I've recently started using `0.17.11`, and am still super happy with the project. Thanks again for all your great work! Everything works well, and makes sense.

I have a few questions, when you have a chance. Forgive me if I am making obvious newbie mistakes :)

--- 

I recently wrote a plugin that would help me to migrate away from Markdown/remark into a different syntax. The process of writing the plugin, migrating the content, and fixing the output was smooth enough, but during this process I had to hack the regular expression in `urlify` to support new filetypes.

My approach was basically to update the `test` in the webpack (1) config, write a couple plugins, reference the new plugins in `webpack.config.babel.js`, and then update `scripts/phenomic.browser.js`. I had to hack `urlify` to stop the extensions from being part of my slugs/uris.

I am not sure if supporting files with extensions other than `md` or `markdown` is supported by `urlify`, but if it is, I couldn't find out how, using `phenomicLoader`. I am also not sure if it is expected that I would have written a new loader, instead of relying on `phenomicLoader`.

In total, I changed three files to get loading the new filetypes working:
1. `webpack.config.babel.js`
2. `scripts/phenomic.browser.js`
3. `node_modules/phenomic/lib/_utils/urlify/index.js`

Obviously, changing number 3 makes me feel not as happy as numbers 1, and 2. Perhaps this is caused by my initial reluctance to write my own loader, instead writing plugins? Is that what is expected?

`node_modules/phenomic/lib/_utils/urlify/index.js` on my project directory went from: 

``` javascript
function urlify(string) {
  // ...
  url = url
  // something/index.md => something
  // something/index.markdown => something
  .replace(/\bindex\.(md|markdown)$/, "")
  // something-else.md => something-else
  // something-else.markdown => something-else
  .replace(/\.(md|markdown)$/, "");
  // ...
}
```

to

``` js
function urlify(string) {
  // ...
  url = url
  // something/index.md => something
  // something/index.markdown => something
  .replace(/\bindex\.(md|markdown|NEW_FILE_EXTENSION)$/, "")
  // something-else.md => something-else
  // something-else.markdown => something-else
  .replace(/\.(md|markdown|NEW_FILE_EXTENSION)$/, "");
  // ...
}
```

it's obviously pretty hacky/kludgy, and it seems to be working fine in the interim, and it's obviously not a long-term solution.

--- 

Hopefully that all makes sense.

My questions are:

1) Am I expected to write a new loader to support new file types/extensions? If yes, then this is not really a problem with phenomic _per se_, but an issue with my initial understanding and approach.
2) Is there any supported way to extend the regular expression in `urlify` without resorting to writing a new loader?

I'm happy to contribute to any solution, if needed, or to try rewriting as a loader and seeing if that works better.

Thanks again!

**Comment by MoOx** [comment]
1) No, phenomic-loader is a core piece so you should not have to write a fork
2) for now no. Well yes actually but it will still be a hack (using webpack alias, you may override any file..)

We need to find a smart solution to this problem. Maybe we can just add most common extensions? Or we can also add an option like "extensionsToHideInUrl" (or a smarter name, you get the idea) to allow people to send us extensions to remove when we use `urlify`.

I guess adding new extensions by hand in the core is probably a better idea. We could start with `(md|markdown|txt|tex|textile|t2t)`? What extension are you using?

**Comment by ezralalonde** [comment]
I'm using AsciiDoc at the moment, so `asciidoc`/`asc`/`adoc` extensions.

A few first guesses at what could be done:

1. Insist people utilize `route` on any documents that they want to go "out-of-the-box" on;
2. Hardcode additional extensions to `urlize`;
3. Allow plugins to override metadata values for `__url`, `__resourceUrl`, `__dataUrl` safely;
4. Just remove all file extensions (everything after last dot?) in `urlize`;
5. Provide some kind of external config for filenames, "extensionsToHideInUrl" as you said.

Obviously they're not mutually exclusive.

My impressions:

1. Gets annoying _really_ fast.
2. Will end up with trivial but annoying maintenance issues for new types? Anything that moves the ability to self-service without hacking dependencies or making a pull request for a new type is better, IMO. Users don't expect certain extensions to be whitelisted.
3. Might be interesting, but sounds quite dangerous to me, who knows nothing about it. I have no clue if this technically possible/desirable currently, or the degree of difficulty to implement.
4. Seems like it's the least involved solution. You could still fallback to using `route` in the rare case this is wrong?
5. Is good. You could use this config easily align the expressions in `webpack.config.babel.js`, `/scripts/phenomic.browser.js`, and `phenomic/lib/_utils/urlify`. If the default config showed `[ "markdown", "md" ]` it could be pretty obvious for users to modify.

I would lean toward (4) or (5).

Ultimately, the "problem" is a little silly, but most people's expectation is that thy can name files however they wish :)

I'm happy to help with a PR if you need. Thank you!

**Comment by MoOx** [comment]
I don't like 5 a lot, cause it's not straight forward.
The problem with 4 is that we will have to maintain a blacklist (eg: .html should not be removed, for example for 404.html or things like that).

I think for now 2 is not so bad?

**Comment by ezralalonde** [comment]
Works for me.

Thanks!

**Comment by bloodyowl** [comment]
Going to be fixed with #925

**Comment by MoOx** [comment]
Probably not relevant anymore with v1.
